libz-rs-sys/src/gz.rs

```diff
@@ -1738,6 +1738,111 @@ pub unsafe extern "C-unwind" fn gzoffset(file: gzFile) -> z_off_t {
     }
 }
 
+/// Compress and write `c`, converted to an unsigned 8-bit char, into `file`.
+///
+/// # Returns
+///
+///  - The value that was written, on success.
+///  - `-1` on error.
+///
+/// # Safety
+///
+/// - `file`, if non-null, must be an open file handle obtained from [`gzopen`] or [`gzdopen`].
+pub unsafe extern "C-unwind" fn gzputc(file: gzFile, c: c_int) -> c_int {
+    let Some(state) = (unsafe { file.cast::<GzState>().as_mut() }) else {
+        return -1;
+    };
+
+    // Check that we're writing and that there's no error.
+    if state.mode != GzMode::GZ_WRITE || state.err != Z_OK {
+        return -1;
+    }
+
+    /* FIXME: Uncomment when seek support is implemented.
+    // Check for seek request.
+    if state.seek {
+        state.seek = false;
+        if gz_zero(state, state.skip) == -1 {
+            return -1;
+        }
+    }
+     */
+
+    // Try writing to input buffer for speed (state.input == null if buffer not initialized).
+    if !state.input.is_null() {
+        if state.stream.avail_in == 0 {
+            state.stream.next_in = state.input;
+        }
+        let have = unsafe {
+            state
+                .stream
+                .next_in
+                .add(state.stream.avail_in as usize)
+                .offset_from(state.input)
+        } as usize;
```

Comment:
We already have this bit of code somewhere else right? Maybe this can be a method on `state`?

Reply:
I just pushed an update where this is encapsulated in a function. I made a small change to `gz_write` to simplify the safety conditions for this function. Let me know if the safety comments make sense.